src/prng/chacha.rs

```diff
@@ -129,66 +127,72 @@ impl ChaChaRng {
         ChaChaRng::from_seed([0; SEED_WORDS*4])
     }
 
-    /// Sets the internal 128-bit ChaCha counter to a user-provided value. This
-    /// permits jumping arbitrarily ahead (or backwards) in the pseudorandom
-    /// stream.
-    ///
-    /// The 128 bits used for the counter overlap with the nonce and smaller
-    /// counter of ChaCha when used as a stream cipher. It is in theory possible
-    /// to use `set_counter` to obtain the conventional ChaCha pseudorandom
-    /// stream associated with a particular nonce. This is not a supported use
-    /// of the RNG, because a nonce set that way is not treated as a constant
-    /// value but still as part of the counter, besides endian issues.
-    ///
-    /// # Examples
-    ///
-    /// ```rust
-    /// use rand::{ChaChaRng, RngCore, SeedableRng};
-    ///
-    /// // Note: Use `FromEntropy` or `ChaChaRng::from_rng()` outside of testing.
-    /// let mut rng1 = ChaChaRng::from_seed([0; 32]);
-    /// let mut rng2 = rng1.clone();
-    ///
-    /// // Skip to round 20. Because every round generates 16 `u32` values, this
-    /// // actually means skipping 320 values.
-    /// for _ in 0..(20*16) { rng1.next_u32(); }
-    /// rng2.set_counter(20, 0);
-    /// assert_eq!(rng1.next_u32(), rng2.next_u32());
-    /// ```
-    pub fn set_counter(&mut self, counter_low: u64, counter_high: u64) {
-        self.0.inner_mut().set_counter(counter_low, counter_high);
-        self.0.reset(); // force recomputation on next use
+    /// Get the offset from the start of the stream, in 32-bit words.
+    /// 
+    /// Since the generated blocks are 16 words (2<sup>4</sup>) long and the
+    /// counter is 64-bits, the offset is a 68-bit number. Sub-word offsets are
+    /// not supported, hence the result can simply be multiplied by 4 to get a
+    /// byte-offset.
```

Comment:
Can you document that the returned position is that of the previous value? And I have a small preference for returning the next value instead. Otherwise you get `u64::MAX` right after initialization.

Maybe also add a test that you can generate a couple of values twice using `get_word_pos` and `set_word_pos`?

Reply:
> you get u64::MAX right after initialization

Sounds like I should fix that!

Reply:
Actually, this is incorrect — `set_word_pos(0)` will set both parts of the counter to zero, then call reset (effectively triggering generation of values on next use). `get_word_pos()` will return zero when the counter is zero and index is high — this relies on the index never being zero, which seems to be an informal contract of the `BlockRng` code. Perhaps we should make that explicit, or test for zero in `get_word_pos`.

Reply:
Oops, I read your code wrong. Yes, it is good :smile:.

> this relies on the index never being zero, which seems to be an informal contract of the `BlockRng` code.

The old ISAAC code, where this comes from, did use a zero index right after seeding. Because a new batch of results was already generated after seeding, and 0 values of those where used.

Your new `generate_and_set` can provide the same, and I see no harm in it.

Reply:
True. Annoying to have to check another thing, but best to be correct. :+1: